diff --git a/common/private/seapp_contexts b/common/private/seapp_contexts
index fc26e985154f85d98301bea5fff03ca4f3e3f093..e0e4addc4d01e30e75635a941c82da7fd3129c24 100644
--- a/common/private/seapp_contexts
+++ b/common/private/seapp_contexts
@@ -1,4 +1,2 @@
 user=_app isPrivApp=true seinfo=platform name=com.alpha.updater domain=updater_app type=app_data_file levelFrom=user
-user=_app seinfo=platform name=com.google.android.flipendo domain=flipendo type=app_data_file levelFrom=all
-user=_app seinfo=platform name=com.google.android.turboadapter domain=turbo_adapter type=app_data_file levelFrom=all
 user=_app isPrivApp=true seinfo=mediashell domain=mediashell_app name=com.google.android.apps.mediashell type=app_data_file levelFrom=all
