diff --git a/core/java/android/provider/Settings.java b/core/java/android/provider/Settings.java
index 3139e3d13118dff8a7fffec2841d834a21cedd4c..7a0494077a250d585ea3dbeb3bda746ccdb32374 100644
--- a/core/java/android/provider/Settings.java
+++ b/core/java/android/provider/Settings.java
@@ -5620,6 +5620,106 @@ public final class Settings {
          */
         public static final String SETTINGS_DASHBOARD_STYLE = "settings_dashboard_style";
 
+        /**
+         * @hide
+         */
+        public static final String ADVANCED_REBOOT_SECURED = "advanced_reboot_secured";
+
+        /**
+         * @hide
+         */
+        public static final String AMBIENT_CUSTOM_IMAGE = "ambient_custom_image";
+
+        /**
+         * @hide
+         */
+        public static final String AMBIENT_TEXT_ALIGNMENT = "ambient_text_alignment";
+
+        /**
+         * @hide
+         */
+        public static final String AMBIENT_TEXT_COLOR = "ambient_text_color";
+
+        /**
+         * @hide
+         */
+        public static final String AMBIENT_TEXT_STRING = "ambient_text_string";
+
+        /**
+         * @hide
+         */
+        public static final String AMBIENT_TEXT_TYPE_COLOR = "ambient_text_type_color";
+
+        /**
+         * @hide
+         */
+        public static final String BATTERY_LIGHT_REALLY_FULL_COLOR = "battery_light_really_full_color";
+
+        /**
+         * @hide
+         */
+        public static final String GLOBAL_ACTION_KEY_ONTHEGO = "global_action_key_onthego";
+
+        /**
+         * @hide
+         */
+        public static final String KEY_APP_SWITCH_DOUBLE_TAP_ACTION = "key_app_switch_double_tap_action";
+
+        /**
+         * @hide
+         */
+        public static final String KEY_ASSIST_DOUBLE_TAP_ACTION = "key_assist_double_tap_action";
+
+        /**
+         * @hide
+         */
+        public static final String KEY_BACK_DOUBLE_TAP_ACTION = "key_back_double_tap_action";
+
+        /**
+         * @hide
+         */
+        public static final String NETWORK_TRAFFIC_AUTOHIDE_THRESHOLD = "network_traffic_autohide_threshold";
+
+        /**
+         * @hide
+         */
+        public static final String NETWORK_TRAFFIC_HIDEARROW = "network_traffic_hidearrow";
+
+        /**
+         * @hide
+         */
+        public static final String NETWORK_TRAFFIC_LOCATION = "network_traffic_location";
+
+        /**
+         * @hide
+         */
+        public static final String NETWORK_TRAFFIC_REFRESH_INTERVAL = "network_traffic_refresh_interval";
+
+        /**
+         * @hide
+         */
+        public static final String QS_BRIGHTNESS_SLIDER_POSITION = "qs_brightness_slider_position";
+
+        /**
+         * @hide
+         */
+        public static final String QS_FOOTER_DATA_USAGE = "qs_footer_data_usage";
+
+        /**
+         * @hide
+         */
+        public static final String UDFPS_COLOR = "udfps_color";
+
+        /**
+         * @hide
+         */
+        public static final String VOLUME_MEDIA_OUTPUT_TOGGLE = "volume_media_output_toggle";
+
+        /**
+         * @hide
+         */
+        public static final String VOWIFI_ICON_STYLE = "vowifi_icon_style";
+
 
 
         /**
@@ -5661,12 +5761,6 @@ public final class Settings {
 
 
 
-        /**
-         * Origins for which browsers should allow geolocation by default.
-         * The value is a space-separated list of origins.
-         */
-        @Readable
-        public static final String ALLOWED_GEOLOCATION_ORIGINS = "allowed_geolocation_origins";
 
 
 
@@ -5914,7 +6008,6 @@ public final class Settings {
      * included the system will filter out any Settings that doesn't appear in this list that
      * otherwise would display.
      */
-    public static final String EXTRA_CHANNEL_FILTER_LIST = "android.provider.extra.CHANNEL_FILTER_LIST";
 
 
 
@@ -6860,22 +6953,13 @@ public final class Settings {
 
 
 
-        /**
-         * ADVANCED_SETTINGS default value.
-         * @hide
-         */
-        public static final int ADVANCED_SETTINGS_DEFAULT = 0;
+
+
 
         /**
-         * If the triple press gesture for toggling accessibility is enabled.
-         * Set to 1 for true and 0 for false.
-         *
-         * This setting is used only internally.
          * @hide
          */
-        public static final String WEAR_ACCESSIBILITY_GESTURE_ENABLED = "wear_accessibility_gesture_enabled";
-
-
+        public static final String CLOCKWORK_HOME_READY = "clockwork_home_ready";
 
         /**
          * Global settings that shouldn't be persisted.
@@ -12678,6 +12762,371 @@ public final class Settings {
         public static void setLocationProviderEnabled(ContentResolver cr,
                 String provider, boolean enabled) {
         }
+
+        /**
+         * Pulse notifications on tilt
+         * @hide
+         */
+        public static final String DOZE_TILT_GESTURE = "doze_tilt_gesture";
+
+
+
+        /**
+         * Pulse notifications on hand wave
+         * @hide
+         */
+        public static final String DOZE_HANDWAVE_GESTURE = "doze_handwave_gesture";
+
+
+
+        /**
+         * Pulse notifications on removal from pocket
+         * @hide
+         */
+        public static final String DOZE_POCKET_GESTURE = "doze_pocket_gesture";
+
+
+
+        /**
+         * Wake up instead of pulsing notifications
+         * @hide
+         */
+        public static final String RAISE_TO_WAKE_GESTURE = "raise_to_wake_gesture";
+
+
+
+        /**
+         * Vibrate when pulsing notifications on gesture
+         * @hide
+         */
+        public static final String DOZE_GESTURE_VIBRATE = "doze_gesture_vibrate";
+
+
+
+        /**
+         * Boolean value whether to link ringtone and notification volume
+         * @hide
+         */
+        public static final String VOLUME_LINK_NOTIFICATION = "volume_link_notification";
+
+
+
+        /**
+         * Whether to show privacy indicator for location
+         * @hide
+         */
+        public static final String ENABLE_LOCATION_PRIVACY_INDICATOR = "enable_location_privacy_indicator";
+
+
+
+        /**
+         * Whether to show privacy indicator for camera
+         * @hide
+         */
+        public static final String ENABLE_CAMERA_PRIVACY_INDICATOR = "enable_camera_privacy_indicator";
+
+
+
+        /**
+         * Whether to show privacy indicator for media projection
+         * @hide
+         */
+        public static final String ENABLE_PROJECTION_PRIVACY_INDICATOR = "enable_projection_privacy_indicator";
+
+
+
+        /**
+         * Enable udfps detection even when screen is off
+         * Default value is 0
+         * @hide
+         */
+        public static final String SCREEN_OFF_UDFPS_ENABLED = "screen_off_udfps_enabled";
+
+
+
+        /**
+         * Select from different navigation bar layouts
+         * @hide
+         */
+        public static final String NAVBAR_LAYOUT_VIEWS = "navbar_layout_views";
+
+
+
+        /**
+         * Inverse navigation bar layout
+         * @hide
+         */
+        public static final String NAVBAR_INVERSE_LAYOUT = "navbar_inverse_layout";
+
+
+
+        /**
+         * Whether to show or hide the arrow for back gesture
+         * @hide
+         */
+        public static final String BACK_GESTURE_ARROW = "back_gesture_arrow";
+
+
+
+        /**
+         * Whether or not to vibrate when back gesture is used
+         * @hide
+         */
+        public static final String BACK_GESTURE_HAPTIC = "back_gesture_haptic";
+
+
+
+        /**
+         * Control whether the process CPU info meter should be shown.
+         * @hide
+         */
+        public static final String SHOW_CPU_OVERLAY = "show_cpu_overlay";
+
+
+
+        /**
+         * Control whether the process FPS info meter should be shown.
+         * @hide
+         */
+        public static final String SHOW_FPS_OVERLAY = "show_fps_overlay";
+
+
+
+        /**
+         * Whether to pulse ambient on new music tracks
+         * @hide
+         */
+        public static final String PULSE_ON_NEW_TRACKS = "pulse_on_new_tracks";
+
+
+
+        /**
+         * Pulse navbar music visualizer
+         * @hide
+         */
+        public static final String NAVBAR_PULSE_ENABLED = "navbar_pulse_enabled";
+
+
+
+        /**
+         * Pulse ambient music visualizer
+         * @hide
+         */
+        public static final String AMBIENT_PULSE_ENABLED = "ambient_pulse_enabled";
+
+
+
+        /**
+         * Pulse lockscreen music visualizer
+         * @hide
+         */
+        public static final String LOCKSCREEN_PULSE_ENABLED = "lockscreen_pulse_enabled";
+
+
+
+        /**
+         * Pulse navbar music visualizer color type
+         * @hide
+         */
+        public static final String PULSE_COLOR_MODE = "pulse_color_mode";
+
+
+
+        /**
+         * Pulse music visualizer user defined color
+         * @hide
+         */
+        public static final String PULSE_COLOR_USER = "pulse_color_user";
+
+
+
+        /**
+         * Pulse lavalamp animation speed
+         * @hide
+         */
+        public static final String PULSE_LAVALAMP_SPEED = "pulse_lavalamp_speed";
+
+
+
+        /**
+         * Pulse renderer implementation
+         * @hide
+         */
+        public static final String PULSE_RENDER_STYLE = "pulse_render_style";
+
+
+
+        /**
+         * Custom Pulse Widths
+         * @hide
+         */
+        public static final String PULSE_CUSTOM_DIMEN = "pulse_custom_dimen";
+
+
+
+        /**
+         * Custom Spacing Between Pulse Bars
+         * @hide
+         */
+        public static final String PULSE_CUSTOM_DIV = "pulse_custom_div";
+
+
+
+        /**
+         * Custom Pulse Block Size
+         * @hide
+         */
+        public static final String PULSE_FILLED_BLOCK_SIZE = "pulse_filled_block_size";
+
+
+
+        /**
+         * Custom Spacing Between Pulse Blocks
+         * @hide
+         */
+        public static final String PULSE_EMPTY_BLOCK_SIZE = "pulse_empty_block_size";
+
+
+
+        /**
+         * Custom Pulse Sanity Levels
+         * @hide
+         */
+        public static final String PULSE_CUSTOM_FUDGE_FACTOR = "pulse_custom_fudge_factor";
+
+
+
+        /**
+         * Pulse Fudge Factor
+         * @hide
+         */
+        public static final String PULSE_SOLID_FUDGE_FACTOR = "pulse_solid_fudge_factor";
+
+
+
+        /**
+         * Pulse Solid units count
+         * @hide
+         */
+        public static final String PULSE_SOLID_UNITS_COUNT = "pulse_solid_units_count";
+
+
+
+        /**
+         * Pulse Solid units opacity
+         * @hide
+         */
+        public static final String PULSE_SOLID_UNITS_OPACITY = "pulse_solid_units_opacity";
+
+
+
+        /**
+         * Pulse Solid units rounded
+         * @hide
+         */
+        public static final String PULSE_SOLID_UNITS_ROUNDED = "pulse_solid_units_rounded";
+
+
+
+        /**
+         * Pulse uses FFT averaging
+         * @hide
+         */
+        public static final String PULSE_SMOOTHING_ENABLED = "pulse_smoothing_enabled";
+
+
+
+        /**
+         * @hide
+         */
+        public static final String KG_CUSTOM_CLOCK_TOP_MARGIN = "kg_custom_clock_top_margin";
+
+
+
+        /**
+         * @hide
+         */
+        public static final String KG_LARGE_CLOCK_TEXT_SIZE = "kg_large_clock_text_size";
+
+
+
+        /**
+         * @hide
+         */
+        public static final String KG_SMALL_CLOCK_TEXT_SIZE = "kg_small_clock_text_size";
+
+
+
+        /**
+         * Whether to show an overlay in the bottom corner of the screen on copying stuff
+         * into the clipboard.
+         * @hide
+         */
+        public static final String SHOW_CLIPBOARD_OVERLAY = "show_clipboard_overlay";
+
+
+
+        /**
+         * Face Unlock Method
+         * @hide
+         */
+        public static final String FACE_UNLOCK_METHOD = "face_unlock_method";
+
+
+
+        /**
+         * Whether to enable DOZE only when charging
+         * @hide
+         */
+        public static final String DOZE_ON_CHARGE = "doze_on_charge";
+
+        /**
+         * @hide
+         */
+        public static final String ADVANCED_REBOOT_SECURED = "advanced_reboot_secured";
+
+
+        /**
+         * @hide
+         */
+        public static final String BATTERY_LIGHT_REALLY_FULL_COLOR = "battery_light_really_full_color";
+
+
+        /**
+         * @hide
+         */
+        public static final String GLOBAL_ACTION_KEY_ONTHEGO = "global_action_key_onthego";
+
+
+        /**
+         * @hide
+         */
+        public static final String NETWORK_TRAFFIC_AUTOHIDE_THRESHOLD = "network_traffic_autohide_threshold";
+
+
+        /**
+         * @hide
+         */
+        public static final String NETWORK_TRAFFIC_HIDEARROW = "network_traffic_hidearrow";
+
+
+        /**
+         * @hide
+         */
+        public static final String NETWORK_TRAFFIC_LOCATION = "network_traffic_location";
+
+
+        /**
+         * @hide
+         */
+        public static final String NETWORK_TRAFFIC_REFRESH_INTERVAL = "network_traffic_refresh_interval";
+
+
+        /**
+         * @hide
+         */
+        public static final String QS_BRIGHTNESS_SLIDER_POSITION = "qs_brightness_slider_position";
+
     }
 
     /**
@@ -19324,7 +19773,30 @@ public final class Settings {
              * @hide
              */
             public static final String WET_MODE_ON = "wet_mode_on";
-        }
+
+        /**
+         * Global settings that shouldn't be persisted.
+         *
+         * @hide
+         */
+        public static final String[] TRANSIENT_SETTINGS = {
+                CLOCKWORK_HOME_READY,
+        };
+
+        /**
+         * The amount of time in milliseconds before wifi is turned off
+         * @hide
+         */
+        public static final String WIFI_OFF_TIMEOUT = "wifi_off_timeout";
+
+
+
+        /**
+         * The amount of time in milliseconds before bluetooth is turned off
+         * @hide
+         */
+        public static final String BLUETOOTH_OFF_TIMEOUT = "bluetooth_off_timeout";
+}
     }
 
     /**
@@ -20298,4 +20770,5 @@ public final class Settings {
         }
         return packages[0];
     }
+
 }
diff --git a/core/java/com/android/internal/app/IBatteryStats.aidl b/core/java/com/android/internal/app/IBatteryStats.aidl
index 629a1b36b9e674c1d6b30030d5bfae176259cc98..c1fa18f0492231c5a12d4dcf84e30884e87cb8a0 100644
--- a/core/java/com/android/internal/app/IBatteryStats.aidl
+++ b/core/java/com/android/internal/app/IBatteryStats.aidl
@@ -189,4 +189,7 @@ interface IBatteryStats {
     void resetBattery(boolean forceUpdate);
     /** Exposed as a test API. */
     void suspendBatteryInput();
+
+    /** {@hide} */
+    void resetStatistics();
 }
diff --git a/core/java/com/android/internal/util/crdroid/ThemeUtils.java b/core/java/com/android/internal/util/crdroid/ThemeUtils.java
index 1ad871daea614aab02c4a1839a3bf0ba7c6b8989..d63a7ab0431c62b3402829d7c03516829aab5ae5 100644
--- a/core/java/com/android/internal/util/crdroid/ThemeUtils.java
+++ b/core/java/com/android/internal/util/crdroid/ThemeUtils.java
@@ -74,21 +74,25 @@ public class ThemeUtils {
     }
 
     public void setOverlayEnabled(String category, String packageName) {
-        final String currentPackageName = getOverlayInfos(category).stream()
+        setOverlayEnabled(category, packageName, "android");
+    }
+
+    public void setOverlayEnabled(String category, String packageName, String target) {
+        final String currentPackageName = getOverlayInfos(category, target).stream()
                 .filter(info -> info.isEnabled())
                 .map(info -> info.packageName)
                 .findFirst()
                 .orElse(null);
 
         try {
-            if ("android".equals(packageName)) {
+            if (target.equals(packageName)) {
                 mOverlayManager.setEnabled(currentPackageName, false, USER_SYSTEM);
             } else {
                 mOverlayManager.setEnabledExclusiveInCategory(packageName,
                         USER_SYSTEM);
             }
 
-            writeSettings(category, packageName, "android".equals(packageName));
+            writeSettings(category, packageName, target.equals(packageName));
 
         } catch (RemoteException e) {
         }
@@ -124,7 +128,7 @@ public class ThemeUtils {
 
     public List<String> getOverlayPackagesForCategory(String category, String target) {
         List<String> overlays = new ArrayList<>();
-        overlays.add("android");
+        overlays.add(target);
         for (OverlayInfo info : getOverlayInfos(category, target)) {
             if (category.equals(info.getCategory())) {
                 overlays.add(info.getPackageName());
diff --git a/core/java/com/android/internal/util/custom/CustomUtils.java b/core/java/com/android/internal/util/custom/CustomUtils.java
index 10dc833b3ebedb676a0ed4112a4d50142ac03316..96f267eb47fd1c087bf547a35f7ad5e1abaccbf5 100644
--- a/core/java/com/android/internal/util/custom/CustomUtils.java
+++ b/core/java/com/android/internal/util/custom/CustomUtils.java
@@ -64,8 +64,8 @@ public class CustomUtils {
 
     public static void showSettingsRestartDialog(Context context) {
         new AlertDialog.Builder(context)
-                .setTitle(R.string.settings_restart_title)
-                .setMessage(R.string.settings_restart_message)
+                .setTitle("Restart required")
+                .setMessage("A restart is required to apply these changes. Do you want to restart now?")
                 .setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
                     public void onClick(DialogInterface dialog, int id) {
                         restartSettings(context);
diff --git a/core/res/res/values/cr_config.xml b/core/res/res/values/cr_config.xml
index a9923b69440f902ffac2c10821175caa3f913c47..92caf8fa006ee157c1f2a2d0c68aceb407454535 100644
--- a/core/res/res/values/cr_config.xml
+++ b/core/res/res/values/cr_config.xml
@@ -145,4 +145,10 @@
   <!-- List of system apps that are allowed to be locked with app lock.
         Use with extreme caution. -->
   <string-array name="config_appLockAllowedSystemApps" translatable="false" />
+
+    <dimen name="status_bar_padding_start">2dp</dimen>
+    <dimen name="status_bar_padding_end">2dp</dimen>
+    <dimen name="status_bar_padding_top">0dp</dimen>
+
+    <integer name="config_notificationsBatteryReallyFullARGB">0xFF00FF00</integer>
 </resources>
diff --git a/core/res/res/values/cr_strings.xml b/core/res/res/values/cr_strings.xml
index 0f13ca276ebb22fd07e5d8af4fc06d299e462bf7..c598dfe6d886298c6317bb6eaa0f4602adb1299a 100644
--- a/core/res/res/values/cr_strings.xml
+++ b/core/res/res/values/cr_strings.xml
@@ -44,4 +44,6 @@
 
     <!-- App lock -->
     <string name="unlock_application">Unlock <xliff:g id="label" example="Telegram">%1$s</xliff:g></string>
+    <string name="settings_restart_title">Restart required</string>
+    <string name="settings_restart_message">A restart is required to apply these changes. Do you want to restart now?</string>
 </resources>
diff --git a/core/res/res/values/cr_symbols.xml b/core/res/res/values/cr_symbols.xml
index 216ed570938e673745600d95ba63d4df612fc144..3d46c590efd3f5bd8979bf4ad9e4b6fa03253404 100644
--- a/core/res/res/values/cr_symbols.xml
+++ b/core/res/res/values/cr_symbols.xml
@@ -128,4 +128,12 @@
     <!-- App lock -->
     <java-symbol type="string" name="unlock_application" />
     <java-symbol type="array" name="config_appLockAllowedSystemApps" />
+
+    <!-- Status bar padding -->
+    <java-symbol type="dimen" name="status_bar_padding_start" />
+    <java-symbol type="dimen" name="status_bar_padding_end" />
+    <java-symbol type="dimen" name="status_bar_padding_top" />
+
+    <!-- Battery light -->
+    <java-symbol type="integer" name="config_notificationsBatteryReallyFullARGB" />
 </resources>
diff --git a/media/java/android/media/AudioManager.java b/media/java/android/media/AudioManager.java
index a38df2c54c4e8f5b3a6bf42bbbbe706a87bc79f3..14aefb1317466e337e1e5eedbe1d26080ac7275a 100644
--- a/media/java/android/media/AudioManager.java
+++ b/media/java/android/media/AudioManager.java
@@ -8780,4 +8780,7 @@ public class AudioManager {
             return mHandler;
         }
     }
+
+    /** @hide */
+    public static final String ACTION_VOLUME_STEPS_CHANGED = "android.media.ACTION_VOLUME_STEPS_CHANGED";
 }
diff --git a/media/java/android/media/IAudioService.aidl b/media/java/android/media/IAudioService.aidl
index 88a0321d34da656168eb7c698b698285b023e72a..8ceda8d00fb10938c31c5d5d4d491bc5ddd995e6 100755
--- a/media/java/android/media/IAudioService.aidl
+++ b/media/java/android/media/IAudioService.aidl
@@ -525,4 +525,8 @@ interface IAudioService {
             boolean handlesvolumeAdjustment);
 
     String getHalVersion();
+
+
+    /** @hide */
+    void setStreamMaxVolume(int streamType, int maxVol);
 }
diff --git a/packages/SettingsLib/src/com/android/settingslib/applications/AppIconCacheManager.java b/packages/SettingsLib/src/com/android/settingslib/applications/AppIconCacheManager.java
index 9dfc8eaac024a8c7a8d6fbfd9b63b6079a7d1c4c..1f3816b2019ce57de75070064087b507a098c625 100644
--- a/packages/SettingsLib/src/com/android/settingslib/applications/AppIconCacheManager.java
+++ b/packages/SettingsLib/src/com/android/settingslib/applications/AppIconCacheManager.java
@@ -99,6 +99,26 @@ public class AppIconCacheManager {
         }
     }
 
+
+    /**
+     * @see android.content.ComponentCallbacks2#onTrimMemory(int)
+     */
+    public void trimMemory(int level) {
+        if (level >= android.content.ComponentCallbacks2.TRIM_MEMORY_BACKGROUND) {
+            // Time to clear everything
+            if (sAppIconCacheManager != null) {
+                sAppIconCacheManager.mDrawableCache.trimToSize(0);
+            }
+        } else if (level >= android.content.ComponentCallbacks2.TRIM_MEMORY_UI_HIDDEN
+                || level == android.content.ComponentCallbacks2.TRIM_MEMORY_RUNNING_CRITICAL) {
+            // tough time but still affordable, clear half of the cache
+            if (sAppIconCacheManager != null) {
+                sAppIconCacheManager.mDrawableCache.trimToSize(
+                        sAppIconCacheManager.mDrawableCache.size() / 2);
+            }
+        }
+    }
+
     private static String getKey(String packageName, int uid) {
         if (packageName == null || uid < 0) {
             return null;
diff --git a/packages/SettingsLib/src/com/android/settingslib/drawer/CategoryKey.java b/packages/SettingsLib/src/com/android/settingslib/drawer/CategoryKey.java
index db224bee2550a3b76de488a5365e2bd3d13d7926..e1f2938097f6bac8b526c4a325a0a72123550749 100644
--- a/packages/SettingsLib/src/com/android/settingslib/drawer/CategoryKey.java
+++ b/packages/SettingsLib/src/com/android/settingslib/drawer/CategoryKey.java
@@ -22,6 +22,7 @@ public final class CategoryKey {
 
     // Activities in this category shows up in Settings homepage.
     public static final String CATEGORY_HOMEPAGE = "com.android.settings.category.ia.homepage";
+    public static final String CATEGORY_ALPHA_SETTINGS = "com.alpha.settings.category.settings";
 
     // Top level category.
     public static final String CATEGORY_NETWORK = "com.android.settings.category.ia.wireless";
diff --git a/packages/SystemUI/Android.bp b/packages/SystemUI/Android.bp
index b4027197344daa90719797322188593cc59773b4..ad1ad5d0edd95b616db1f77aed17d4e7b8301964 100644
--- a/packages/SystemUI/Android.bp
+++ b/packages/SystemUI/Android.bp
@@ -172,6 +172,7 @@ android_library {
         "SystemUI-tags",
         "SystemUI-proto",
         "monet",
+        "vendor.goodix.hardware.biometrics.fingerprint-V2.1-java",
         "dagger2",
         "jsr330",
         "lottie",
diff --git a/packages/SystemUI/src/com/android/keyguard/KeyguardPasswordViewController.java b/packages/SystemUI/src/com/android/keyguard/KeyguardPasswordViewController.java
index 6a0db1906706d62502dbcb1b8cab909b74e032bc..e77ee036cf77753df52f021878b291e341d5100a 100644
--- a/packages/SystemUI/src/com/android/keyguard/KeyguardPasswordViewController.java
+++ b/packages/SystemUI/src/com/android/keyguard/KeyguardPasswordViewController.java
@@ -17,7 +17,9 @@
 package com.android.keyguard;
 
 import android.content.res.Resources;
+import android.os.AsyncTask;
 import android.os.UserHandle;
+import android.provider.Settings;
 import android.text.Editable;
 import android.text.InputType;
 import android.text.TextUtils;
@@ -36,6 +38,8 @@ import android.widget.TextView.OnEditorActionListener;
 
 import com.android.internal.util.LatencyTracker;
 import com.android.internal.widget.LockPatternUtils;
+import com.android.internal.widget.LockPatternUtils.RequestThrottledException;
+import com.android.internal.widget.LockscreenCredential;
 import com.android.keyguard.KeyguardSecurityModel.SecurityMode;
 import com.android.systemui.R;
 import com.android.systemui.classifier.FalsingCollector;
@@ -57,6 +61,8 @@ public class KeyguardPasswordViewController
     private EditText mPasswordEntry;
     private ImageView mSwitchImeButton;
     private boolean mPaused;
+    private boolean mQuickUnlock;
+    private final int userId = KeyguardUpdateMonitor.getCurrentUser();
 
     private final OnEditorActionListener mOnEditorActionListener = (v, actionId, event) -> {
         // Check if this was the result of hitting the enter key
@@ -78,6 +84,9 @@ public class KeyguardPasswordViewController
         @Override
         public void beforeTextChanged(CharSequence s, int start, int count, int after) {
             mKeyguardSecurityCallback.userActivity();
+            mQuickUnlock = Settings.System.getIntForUser(getContext().getContentResolver(),
+                    Settings.System.LOCKSCREEN_QUICK_UNLOCK_CONTROL, 0,
+                    UserHandle.USER_CURRENT) == 1;
         }
 
         @Override
@@ -88,7 +97,7 @@ public class KeyguardPasswordViewController
         public void afterTextChanged(Editable s) {
             if (!TextUtils.isEmpty(s)) {
                 onUserInput();
-                if (quickUnlock) {
+                if (mQuickUnlock) {
                     if (s.length() == mLockPatternUtils.getCredentialLength(userId)) {
                         validateQuickUnlock(mLockPatternUtils, mView.getEnteredCredential(), userId);
                     }
diff --git a/packages/SystemUI/src/com/android/keyguard/KeyguardPinViewController.java b/packages/SystemUI/src/com/android/keyguard/KeyguardPinViewController.java
index eefb5b0972f73acb21824d8047fa2ead2fed95ec..c27e19bca20692f5b7b0542580d50f5ab01edc34 100644
--- a/packages/SystemUI/src/com/android/keyguard/KeyguardPinViewController.java
+++ b/packages/SystemUI/src/com/android/keyguard/KeyguardPinViewController.java
@@ -16,21 +16,40 @@
 
 package com.android.keyguard;
 
+import android.os.AsyncTask;
+import android.os.UserHandle;
+import android.provider.Settings;
 import android.view.View;
 
+import androidx.constraintlayout.widget.ConstraintLayout;
+
 import com.android.internal.util.LatencyTracker;
 import com.android.internal.widget.LockPatternUtils;
+import com.android.internal.widget.LockPatternUtils.RequestThrottledException;
+import com.android.internal.widget.LockscreenCredential;
+import com.android.keyguard.PasswordTextView.QuickUnlockListener;
 import com.android.keyguard.KeyguardSecurityModel.SecurityMode;
 import com.android.systemui.R;
 import com.android.systemui.classifier.FalsingCollector;
 import com.android.systemui.statusbar.policy.DevicePostureController;
 
+import lineageos.providers.LineageSettings;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
 public class KeyguardPinViewController
         extends KeyguardPinBasedInputViewController<KeyguardPINView> {
     private final KeyguardUpdateMonitor mKeyguardUpdateMonitor;
     private final DevicePostureController mPostureController;
     private final DevicePostureController.Callback mPostureCallback = posture ->
             mView.onDevicePostureChanged(posture);
+    private final int userId = KeyguardUpdateMonitor.getCurrentUser();
+    private final KeyguardSecurityCallback mKeyguardSecurityCallback;
+    private final List<Integer> mNumbers =
+            Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9, 0);
 
     protected KeyguardPinViewController(KeyguardPINView view,
             KeyguardUpdateMonitor keyguardUpdateMonitor,
@@ -46,6 +65,7 @@ public class KeyguardPinViewController
                 emergencyButtonController, falsingCollector);
         mKeyguardUpdateMonitor = keyguardUpdateMonitor;
         mPostureController = postureController;
+        mKeyguardSecurityCallback = keyguardSecurityCallback;
     }
 
     @Override
@@ -65,7 +85,7 @@ public class KeyguardPinViewController
                 UserHandle.USER_CURRENT) == 1;
 
         if (scramblePin) {
-            Collections.shuffle(sNumbers);
+            Collections.shuffle(mNumbers);
             // get all children who are NumPadKey's
             ConstraintLayout container = (ConstraintLayout) mView.findViewById(R.id.pin_container);
 
@@ -78,9 +98,9 @@ public class KeyguardPinViewController
             }
 
             // reset the digits in the views
-            for (int i = 0; i < sNumbers.size(); i++) {
+            for (int i = 0; i < mNumbers.size(); i++) {
                 NumPadKey view = views.get(i);
-                view.setDigit(sNumbers.get(i));
+                view.setDigit(mNumbers.get(i));
             }
         }
 
diff --git a/packages/SystemUI/src/com/android/keyguard/PasswordTextView.java b/packages/SystemUI/src/com/android/keyguard/PasswordTextView.java
index 4881c9144269afc40c7282eb4b5a5c1179b087a2..cfa33d68d63055c27c56a029a4a7dd1b044f5042 100644
--- a/packages/SystemUI/src/com/android/keyguard/PasswordTextView.java
+++ b/packages/SystemUI/src/com/android/keyguard/PasswordTextView.java
@@ -99,11 +99,20 @@ public class PasswordTextView extends View {
     private Interpolator mFastOutSlowInInterpolator;
     private boolean mShowPassword = true;
     private UserActivityListener mUserActivityListener;
+    private QuickUnlockListener mQuickUnlockListener;
 
     public interface UserActivityListener {
         void onUserActivity();
     }
 
+    public interface QuickUnlockListener {
+        void onValidateQuickUnlock(String password);
+    }
+
+    public void setQuickUnlockListener(QuickUnlockListener listener) {
+        mQuickUnlockListener = listener;
+    }
+
     public PasswordTextView(Context context) {
         this(context, null);
     }
@@ -260,6 +269,10 @@ public class PasswordTextView extends View {
         }
         userActivity();
         sendAccessibilityEventTypeViewTextChanged(textbefore, textbefore.length(), 0, 1);
+
+        if (mQuickUnlockListener != null) {
+            mQuickUnlockListener.onValidateQuickUnlock(mText);
+        }
     }
 
     public void setUserActivityListener(UserActivityListener userActivitiListener) {
diff --git a/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsView.kt b/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsView.kt
index 8d075aa0b04fb8cc056e1529a6fdb3f59ea78e86..52fca72433280278e932569a5151997c84cc8652 100644
--- a/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsView.kt
+++ b/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsView.kt
@@ -22,6 +22,7 @@ import android.graphics.Paint
 import android.graphics.PointF
 import android.graphics.Rect
 import android.graphics.RectF
+import android.os.FileObserver
 import android.util.AttributeSet
 import android.util.Log
 import android.view.MotionEvent
@@ -33,6 +34,8 @@ import android.widget.FrameLayout
 import android.graphics.PixelFormat
 import com.android.systemui.R
 import com.android.systemui.doze.DozeReceiver
+import java.io.File
+import vendor.goodix.hardware.biometrics.fingerprint.V2_1.IGoodixFingerprintDaemon
 
 private const val TAG = "UdfpsView"
 
@@ -207,6 +210,25 @@ Log.d("PHH", "Surface destroyed!")
             !(animationViewController?.shouldPauseAuth() ?: false)
     }
 
+    private fun goodixCmd(id: Int) {
+        try {
+            IGoodixFingerprintDaemon.getService()?.sendCommand(
+                id,
+                ArrayList()
+            ) { returnCode, _ ->
+                Log.e("PHH-Enroll", "Goodix send command returned code $returnCode")
+            }
+        } catch (t: Throwable) {
+            Log.d("PHH-Enroll", "Failed sending Goodix command $id", t)
+        }
+    }
+
+    private val asusGhbmOnAchieved = "/sys/class/drm/ghbm_on_achieved"
+    private val hasAsusGhbm = File(asusGhbmOnAchieved).exists()
+    private val samsungActualMaskBrightness = "/sys/class/lcd/panel/actual_mask_brightness"
+    private val hasSamsungMask = File(samsungActualMaskBrightness).exists()
+    private var fodFileObserver: FileObserver? = null
+
     fun configureDisplay(onDisplayConfigured: Runnable) {
         isDisplayConfigured = true
         animationViewController?.onDisplayConfiguring()
@@ -217,6 +239,28 @@ Log.d("PHH", "Surface destroyed!")
             gView.startGhbmIllumination(onDisplayConfigured)
         } else {
             doIlluminate(null /* surface */, onDisplayConfigured)        }
+
+        mySurfaceView.visibility = VISIBLE
+        Log.d("PHH", "setting surface visible!")
+
+        val brightnessFile = File("/sys/class/backlight/panel/brightness").let {
+            if (it.exists()) it else File("/sys/class/backlight/panel0-backlight/brightness")
+        }
+        val maxBrightnessFile = File("/sys/class/backlight/panel/max_brightness").let {
+            if (it.exists()) it else File("/sys/class/backlight/panel0-backlight/max_brightness")
+        }
+        if ((hasAsusGhbm || hasSamsungMask)
+                && brightnessFile.exists() && maxBrightnessFile.exists()) {
+            val brightness = brightnessFile.readText().toDouble()
+            val maxBrightness = maxBrightnessFile.readText().toDouble()
+            val dim = 1.0 - Math.pow(brightness / maxBrightness, 1 / 2.3)
+            Log.d("PHH-Enroll", "Brightness is $brightness / $maxBrightness, setting dim to $dim")
+            dimUpdate(dim.toFloat())
+        }
+
+        if (android.os.SystemProperties.get("ro.vendor.build.fingerprint").contains("ASUS")) {
+            goodixCmd(200001)
+        }
     }
 
     private fun doIlluminate(surface: Surface?, onDisplayConfigured: Runnable?) {
@@ -238,5 +282,35 @@ Log.d("PHH", "Surface destroyed!")
             view.visibility = INVISIBLE
         }
         mUdfpsDisplayMode?.disable(null /* onDisabled */)
+
+        val observedPath = when {
+            hasAsusGhbm -> asusGhbmOnAchieved
+            hasSamsungMask -> samsungActualMaskBrightness
+            else -> null
+        }
+        if (observedPath != null) {
+            fodFileObserver = object : FileObserver(observedPath, MODIFY) {
+                override fun onEvent(event: Int, path: String?) {
+                    Log.d("PHH-Enroll", "FOD brightness event on $observedPath")
+                    try {
+                        if (File(observedPath).readText().toInt() == 0) {
+                            dimUpdate(0.0f)
+                            stopWatching()
+                            fodFileObserver = null
+                        }
+                    } catch (e: Exception) {
+                        Log.d("PHH-Enroll", "Failed dim update off", e)
+                    }
+                }
+            }.also { it.startWatching() }
+        } else {
+            dimUpdate(0.0f)
+        }
+
+        mySurfaceView.visibility = INVISIBLE
+        Log.d("PHH", "setting surface invisible!")
+        if (android.os.SystemProperties.get("ro.vendor.build.fingerprint").contains("ASUS")) {
+            goodixCmd(200003)
+        }
     }
 }
diff --git a/services/core/Android.bp b/services/core/Android.bp
index a29865214e6d8f585baf976b81fe8e24f9d07ff2..c3dd2ce57fae9aa6daba3bdea49dcdae28313138 100644
--- a/services/core/Android.bp
+++ b/services/core/Android.bp
@@ -177,6 +177,13 @@ java_library_static {
         "motorola.hardware.health-V1.0-java",
         "overlayable_policy_aidl-java",
         "com.android.sysprop.watchdog",
+        // HIDL
+        "vendor.samsung.hardware.sysinput-V1.2-java",
+        "vendor.goodix.hardware.biometrics.fingerprint-V2.1-java",
+        "vendor.samsung.hardware.biometrics.fingerprint-V3.0-java",
+        "vendor.oplus.hardware.biometrics.fingerprint-V2.1-java",
+        "vendor.oppo.hardware.biometrics.fingerprint-V2.1-java",
+        "vendor.mediatek.hardware.mtkpower-V1.1-java",
         "faceunlock_framework",
         "kotlin-stdlib",
         "kotlinx_coroutines_android",
diff --git a/services/core/java/com/android/server/am/BatteryStatsService.java b/services/core/java/com/android/server/am/BatteryStatsService.java
index 2a2e7839288fa844c7fbe588ef61b130913e49fe..9521fa5a9cdbeafeb8e83396f7675a25de49af84 100644
--- a/services/core/java/com/android/server/am/BatteryStatsService.java
+++ b/services/core/java/com/android/server/am/BatteryStatsService.java
@@ -3042,4 +3042,13 @@ public final class BatteryStatsService extends IBatteryStats.Stub
     public void suspendBatteryInput() {
         mBatteryManagerInternal.suspendBatteryInput();
     }
+
+
+    @Override
+    public void resetStatistics() {
+        synchronized (mStats) {
+            mStats.resetAllStatsCmdLocked();
+            mBatteryUsageStatsStore.removeAllSnapshots();
+        }
+    }
 }
diff --git a/services/core/java/com/android/server/audio/AudioService.java b/services/core/java/com/android/server/audio/AudioService.java
index 3ad569da58d44769a096fb592ce9eb198721c08c..3c64ab47831d094aefd39abc182f7eb8544d525d 100644
--- a/services/core/java/com/android/server/audio/AudioService.java
+++ b/services/core/java/com/android/server/audio/AudioService.java
@@ -4817,6 +4817,11 @@ public class AudioService extends IAudioService.Stub
     }
 
     /** @see AudioManager#getStreamVolume(int) */
+
+    protected static void setMaxStreamVolume(int streamType, int maxVol) {
+        MAX_STREAM_VOLUME[streamType] = maxVol;
+    }
+
     public int getStreamVolume(int streamType) {
         ensureValidStreamType(streamType);
         int device = getDeviceForStream(streamType);
@@ -4875,6 +4880,14 @@ public class AudioService extends IAudioService.Stub
         return (mStreamStates[streamType].getMaxIndex() + 5) / 10;
     }
 
+
+    /** @see AudioManager#setStreamMaxVolume(int,int) */
+    public void setStreamMaxVolume(int streamType, int maxVol) {
+        ensureValidStreamType(streamType);
+        mStreamStates[streamType].setMaxIndex(maxVol);
+        setMaxStreamVolume(streamType, maxVol);
+    }
+
     /** @see AudioManager#getStreamMinVolumeInt(int)
      * Part of service interface, check permissions here */
     public int getStreamMinVolume(int streamType) {
@@ -8341,6 +8354,16 @@ public class AudioService extends IAudioService.Stub
             return changed;
         }
 
+        public void setMaxIndex(int maxVol) {
+            mIndexMax = maxVol;
+            AudioSystem.initStreamVolume(mStreamType, 0, mIndexMax);
+            mIndexMax = maxVol;
+            mIndexMax *= 10;
+            // Volume steps changed, fire the intent.
+            Intent intent = new Intent(AudioManager.ACTION_VOLUME_STEPS_CHANGED);
+            sendBroadcastToAll(intent);
+        }
+
         public int getIndex(int device) {
             synchronized (VolumeStreamState.class) {
                 int index = mIndexMap.get(device, -1);
diff --git a/services/core/java/com/android/server/biometrics/AuthService.java b/services/core/java/com/android/server/biometrics/AuthService.java
index fffb445895a5ecba555cfcd8cbd68cde1d189a6b..64bc42454ec00a276aea1f9b2f04e9860d9368fa 100644
--- a/services/core/java/com/android/server/biometrics/AuthService.java
+++ b/services/core/java/com/android/server/biometrics/AuthService.java
@@ -38,6 +38,7 @@ import android.annotation.Nullable;
 import android.app.AppOpsManager;
 import android.content.Context;
 import android.content.pm.PackageManager;
+import android.graphics.Point;
 import android.hardware.biometrics.BiometricAuthenticator;
 import android.hardware.biometrics.BiometricManager;
 import android.hardware.biometrics.ComponentInfoInternal;
@@ -57,15 +58,18 @@ import android.hardware.face.IFaceService;
 import android.hardware.fingerprint.FingerprintSensorProperties;
 import android.hardware.fingerprint.FingerprintSensorPropertiesInternal;
 import android.hardware.fingerprint.IFingerprintService;
+import android.hardware.display.DisplayManager;
 import android.hardware.iris.IIrisService;
 import android.os.Binder;
 import android.os.Build;
+import android.os.FileObserver;
 import android.os.IBinder;
 import android.os.RemoteException;
 import android.os.ServiceManager;
 import android.os.SystemProperties;
 import android.os.UserHandle;
 import android.provider.Settings;
+import android.util.DisplayMetrics;
 import android.util.Slog;
 
 import com.android.internal.R;
@@ -73,10 +77,18 @@ import com.android.internal.annotations.VisibleForTesting;
 import com.android.internal.util.ArrayUtils;
 import com.android.server.SystemService;
 
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.PrintWriter;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 
+import vendor.goodix.hardware.biometrics.fingerprint.V2_1.IGoodixFingerprintDaemon;
+import vendor.samsung.hardware.biometrics.fingerprint.V3_0.ISehBiometricsFingerprint;
+import vendor.samsung.hardware.sysinput.V1_0.ISehSysInputDev;
+
 /**
  * System service that provides an interface for authenticating with biometrics and
  * PIN/pattern/password to BiometricPrompt and lock screen.
@@ -95,6 +107,9 @@ public class AuthService extends SystemService {
     @VisibleForTesting
     final IAuthService.Stub mImpl;
 
+    private FileObserver mFodFileObserver;
+    private ISehBiometricsFingerprint mSamsungFingerprint;
+
     /**
      * Class for injecting dependencies into AuthService.
      * TODO(b/141025588): Replace with a dependency injection framework (e.g. Guice, Dagger).
@@ -633,11 +648,11 @@ public class AuthService extends SystemService {
             final int firstApiLevel = SystemProperties.getInt(SYSPROP_FIRST_API_LEVEL, 0);
             final int apiLevel = SystemProperties.getInt(SYSPROP_API_LEVEL, firstApiLevel);
             String[] configStrings = mInjector.getConfiguration(getContext());
-            if (configStrings.length == 0 && apiLevel == Build.VERSION_CODES.R) {
+            if (configStrings.length == 0) {
                 // For backwards compatibility with R where biometrics could work without being
                 // configured in config_biometric_sensors. In the absence of a vendor provided
                 // configuration, we assume the weakest biometric strength (i.e. convenience).
-                Slog.w(TAG, "Found R vendor partition without config_biometric_sensors");
+                Slog.w(TAG, "Found vendor partition without config_biometric_sensors");
                 configStrings = generateRSdkCompatibleConfiguration();
             }
             hidlConfigs = new SensorConfig[configStrings.length];
@@ -652,6 +667,69 @@ public class AuthService extends SystemService {
         registerAuthenticators(hidlConfigs);
 
         mInjector.publishBinderService(this, mImpl);
+        try {
+            mSamsungFingerprint = ISehBiometricsFingerprint.getService();
+            Slog.d("PHH", "Got Samsung fingerprint HAL");
+        } catch (Exception e) {
+            Slog.d("PHH", "Failed getting Samsung fingerprint HAL", e);
+        }
+        if (samsungHasCmd("fod_enable") && mSamsungFingerprint != null) {
+            samsungCmd("fod_enable,1,1,0");
+            final String actualMaskBrightnessPath =
+                    "/sys/class/lcd/panel/actual_mask_brightness";
+            Slog.d("PHH-Enroll", "Reading actual brightness file gives "
+                    + readFile(actualMaskBrightnessPath));
+            mFodFileObserver = new FileObserver(actualMaskBrightnessPath, FileObserver.MODIFY) {
+                @Override
+                public void onEvent(int event, String path) {
+                    final String actualMask = readFile(actualMaskBrightnessPath);
+                    try {
+                        mSamsungFingerprint = ISehBiometricsFingerprint.getService();
+                    } catch (Exception ignored) {
+                    }
+                    Slog.d("PHH-Enroll", "New actual mask brightness is " + actualMask);
+                    try {
+                        final int eventReq = "0".equals(actualMask) ? 1 : 2;
+                        if (mSamsungFingerprint != null) {
+                            mSamsungFingerprint.sehRequest(22 /* SEM_FINGER_STATE */, eventReq,
+                                    new ArrayList<>(), (retval, out) -> { });
+                        }
+                    } catch (Exception e) {
+                        Slog.d("PHH-Enroll", "Failed setting Samsung mask event", e);
+                    }
+                }
+            };
+            mFodFileObserver.startWatching();
+        }
+
+        final String asusGhbmOnAchieved = "/sys/class/drm/ghbm_on_achieved";
+        if (new File(asusGhbmOnAchieved).exists()) {
+            mFodFileObserver = new FileObserver(asusGhbmOnAchieved, FileObserver.MODIFY) {
+                private boolean mWasOn;
+
+                @Override
+                public void onEvent(int event, String path) {
+                    final String spotOn = readFile(asusGhbmOnAchieved);
+                    if ("1".equals(spotOn)) {
+                        if (!mWasOn) {
+                            try {
+                                final IGoodixFingerprintDaemon goodixDaemon =
+                                        IGoodixFingerprintDaemon.getService();
+                                goodixDaemon.sendCommand(200002, new ArrayList<>(),
+                                        (returnCode, resultData) -> Slog.e(TAG,
+                                                "Goodix UI-ready command returned " + returnCode));
+                            } catch (Throwable t) {
+                                Slog.d("PHH-Enroll", "Failed sending Goodix UI-ready command", t);
+                            }
+                        }
+                        mWasOn = true;
+                    } else {
+                        mWasOn = false;
+                    }
+                }
+            };
+            mFodFileObserver.startWatching();
+        }
     }
 
     /**
@@ -668,7 +746,7 @@ public class AuthService extends SystemService {
         if (pm.hasSystemFeature(PackageManager.FEATURE_FACE)) {
             modalities.add(String.valueOf(BiometricAuthenticator.TYPE_FACE));
         }
-        final String strength = String.valueOf(Authenticators.BIOMETRIC_CONVENIENCE);
+        final String strength = String.valueOf(Authenticators.BIOMETRIC_STRONG);
         final String[] configStrings = new String[modalities.size()];
         for (int i = 0; i < modalities.size(); ++i) {
             final String id = String.valueOf(i);
@@ -775,18 +853,86 @@ public class AuthService extends SystemService {
                 ? modality : (modality & ~BiometricAuthenticator.TYPE_CREDENTIAL);
     }
 
+    private int[] dynamicUdfpsProps() {
+        final DisplayManager displayManager =
+                (DisplayManager) getContext().getSystemService(Context.DISPLAY_SERVICE);
+        final Point displayRealSize = new Point();
+        final DisplayMetrics displayMetrics = new DisplayMetrics();
+        displayManager.getDisplay(0).getRealSize(displayRealSize);
+        displayManager.getDisplay(0).getMetrics(displayMetrics);
+
+        final String samsungPosition = readFile("/sys/class/fingerprint/fingerprint/position");
+        if (samsungPosition != null) {
+            try {
+                final ISehSysInputDev sysInput = ISehSysInputDev.getService();
+                sysInput.getTspFodInformation(0,
+                        (result, information) -> Slog.d("PHH-Enroll",
+                                "TspFod information " + result + ", " + information));
+                sysInput.getTspFodPosition(0,
+                        (result, position) -> Slog.d("PHH-Enroll",
+                                "TspFod position " + result + ", " + position));
+            } catch (Throwable t) {
+                Slog.d("PHH-Enroll", "Failed reading Samsung TSP FOD information", t);
+            }
+
+            try {
+                final String[] fodPositionArray = samsungPosition.split(",");
+                final float bottomInch = Float.parseFloat(fodPositionArray[0]) * 0.0393700787f;
+                final float areaSizeInch = Float.parseFloat(fodPositionArray[5]) * 0.0393700787f;
+                final int areaSizePx = (int) (areaSizeInch * displayMetrics.ydpi);
+                final float sensorX = displayRealSize.x / 2.0f;
+                final float sensorY = displayRealSize.y - (bottomInch * displayMetrics.ydpi)
+                        - (areaSizeInch * displayMetrics.ydpi / 2.0f);
+
+                samsungCmd(String.format("fod_rect,%d,%d,%d,%d",
+                        (int) (sensorX - areaSizePx / 4.0f),
+                        (int) (sensorY - areaSizePx / 4.0f),
+                        (int) (sensorX + areaSizePx / 4.0f),
+                        (int) (sensorY + areaSizePx / 4.0f)));
+
+                try {
+                    mSamsungFingerprint = ISehBiometricsFingerprint.getService();
+                    mSamsungFingerprint.sehRequest(6, 0, new ArrayList<>(),
+                            (retval, out) -> Slog.d("PHH-Enroll",
+                                    "Samsung sensor status result " + retval));
+                    mSamsungFingerprint.sehRequest(32, 0, new ArrayList<>(),
+                            (retval, out) -> Slog.d("PHH-Enroll",
+                                    "Samsung sensor brightness result " + retval));
+                } catch (Exception e) {
+                    Slog.d("PHH-Enroll", "Failed querying Samsung fingerprint HAL", e);
+                }
+
+                return new int[] {(int) sensorX, (int) sensorY, areaSizePx / 2};
+            } catch (RuntimeException e) {
+                Slog.d("PHH-Enroll", "Failed parsing Samsung FOD position", e);
+            }
+        }
+
+        if (SystemProperties.get("ro.vendor.build.fingerprint").contains("ASUS_I006D")) {
+            return new int[] {displayRealSize.x / 2, 1741, 110};
+        }
+        return new int[0];
+    }
 
     private FingerprintSensorPropertiesInternal getHidlFingerprintSensorProps(int sensorId,
             @BiometricManager.Authenticators.Types int strength) {
         // The existence of config_udfps_sensor_props indicates that the sensor is UDFPS.
-        final int[] udfpsProps = getContext().getResources().getIntArray(
+        int[] udfpsProps = getContext().getResources().getIntArray(
                 com.android.internal.R.array.config_udfps_sensor_props);
 
+        if (ArrayUtils.isEmpty(udfpsProps)) {
+            udfpsProps = dynamicUdfpsProps();
+        }
+
         // Non-empty workaroundLocations indicates that the sensor is SFPS.
         final List<SensorLocationInternal> workaroundLocations =
                 getWorkaroundSensorProps(getContext());
 
         final boolean isUdfps = !ArrayUtils.isEmpty(udfpsProps);
+        if (isUdfps) {
+            Slog.d("PHH-Enroll", "Got UDFPS info " + udfpsProps[0] + ", "
+                    + udfpsProps[1] + ", " + udfpsProps[2]);
+        }
 
         // config_is_powerbutton_fps indicates whether device has a power button fingerprint sensor.
         final boolean isPowerbuttonFps = getContext().getResources().getBoolean(
@@ -854,4 +1000,51 @@ public class AuthService extends SystemService {
                 componentInfo, resetLockoutRequiresHardwareAuthToken,
                 resetLockoutRequiresChallenge);
     }
+
+    private static boolean samsungHasCmd(String cmd) {
+        try {
+            final File file = new File("/sys/devices/virtual/sec/tsp/cmd_list");
+            if (!file.exists()) {
+                return false;
+            }
+            try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
+                String line;
+                while ((line = reader.readLine()) != null) {
+                    if (line.equals(cmd)) {
+                        return true;
+                    }
+                }
+            }
+        } catch (Exception e) {
+            Slog.d("PHH", "Failed reading Samsung command list", e);
+        }
+        return false;
+    }
+
+    private static void samsungCmd(String cmd) {
+        try {
+            writeFile("/sys/devices/virtual/sec/tsp/cmd", cmd);
+            final String status = readFile("/sys/devices/virtual/sec/tsp/cmd_status");
+            final String result = readFile("/sys/devices/virtual/sec/tsp/cmd_result");
+            Slog.d("PHH", "Sending command " + cmd + " returned " + result + ":" + status);
+        } catch (Exception e) {
+            Slog.d("PHH", "Failed sending Samsung command " + cmd, e);
+        }
+    }
+
+    private static void writeFile(String path, String value) {
+        try (PrintWriter writer = new PrintWriter(path, "UTF-8")) {
+            writer.println(value);
+        } catch (Exception e) {
+            Slog.d("PHH", "Failed writing " + value + " to " + path, e);
+        }
+    }
+
+    private static String readFile(String path) {
+        try (BufferedReader reader = new BufferedReader(new FileReader(new File(path)))) {
+            return reader.readLine();
+        } catch (Exception e) {
+            return null;
+        }
+    }
 }
diff --git a/services/core/java/com/android/server/display/LocalDisplayAdapter.java b/services/core/java/com/android/server/display/LocalDisplayAdapter.java
index 58a182a61e44e1a8ea85fcdb10b6306b249394bc..c06b2117182b711d63dd95bd9f08e4163d1a8edd 100644
--- a/services/core/java/com/android/server/display/LocalDisplayAdapter.java
+++ b/services/core/java/com/android/server/display/LocalDisplayAdapter.java
@@ -31,6 +31,7 @@ import android.os.PowerManager;
 import android.os.SystemProperties;
 import android.os.Trace;
 import android.util.DisplayUtils;
+import android.util.Log;
 import android.util.LongSparseArray;
 import android.util.Slog;
 import android.util.SparseArray;
@@ -55,6 +56,8 @@ import java.util.Collections;
 import java.util.List;
 import java.util.Objects;
 
+import vendor.samsung.hardware.sysinput.V1_1.ISehSysInputDev;
+
 /**
  * A display adapter for the local displays managed by SurfaceFlinger.
  * <p>
@@ -228,6 +231,8 @@ final class LocalDisplayAdapter extends DisplayAdapter {
 
         private DisplayEventReceiver.FrameRateOverride[] mFrameRateOverrides =
                 new DisplayEventReceiver.FrameRateOverride[0];
+        private boolean triedSamsungHal = false;
+        private ISehSysInputDev samsungSysinput = null;
 
         LocalDisplayDevice(IBinder displayToken, long physicalDisplayId,
                 SurfaceControl.StaticDisplayInfo staticDisplayInfo,
@@ -815,16 +820,42 @@ final class LocalDisplayAdapter extends DisplayAdapter {
                             }
                             mSidekickActive = false;
                         }
+                        if (!triedSamsungHal) {
+                            triedSamsungHal = true;
+                            try {
+                                samsungSysinput = ISehSysInputDev.getService();
+                            } catch(Throwable t) {}
+                        }
                         final int mode = getPowerModeForState(state);
                         Trace.traceBegin(Trace.TRACE_TAG_POWER, "setDisplayState("
                                 + "id=" + physicalDisplayId
                                 + ", state=" + Display.stateToString(state) + ")");
+
+                        if (samsungSysinput != null) {
+                            try {
+                                Log.d("PHH", "setTspEnable 1, " + state + ", true");
+                                samsungSysinput.setTspEnable(1, state, true);
+                            } catch(Throwable t) {
+                                Log.d("PHH", "Failed settings tsp enable", t);
+                            }
+                        }
+
                         try {
                             mSurfaceControlProxy.setDisplayPowerMode(token, mode);
                             Trace.traceCounter(Trace.TRACE_TAG_POWER, "DisplayPowerMode", mode);
                         } finally {
                             Trace.traceEnd(Trace.TRACE_TAG_POWER);
                         }
+
+                        if (samsungSysinput != null) {
+                            try {
+                                Log.d("PHH", "setTspEnable 1, " + state + ", false");
+                                samsungSysinput.setTspEnable(1, state, false);
+                            } catch(Throwable t) {
+                                Log.d("PHH", "Failed settings tsp enable", t);
+                            }
+                        }
+
                         setCommittedState(state);
                         // If we're entering a suspended (but not OFF) power state and we
                         // have a sidekick available, tell it now that it can take control.
diff --git a/services/core/java/com/android/server/wm/DisplayPolicy.java b/services/core/java/com/android/server/wm/DisplayPolicy.java
index 1468360b37b76f326b65389b9c7ff1225404db28..b2436d9159e1249d185652d8fe22c1fc9794512d 100644
--- a/services/core/java/com/android/server/wm/DisplayPolicy.java
+++ b/services/core/java/com/android/server/wm/DisplayPolicy.java
@@ -222,6 +222,9 @@ public class DisplayPolicy {
     private boolean mCanSystemBarsBeShownByUser;
     private boolean mNavButtonForcedVisible;
 
+    private vendor.mediatek.hardware.mtkpower.V1_1.IMtkPerf mMtkPerf;
+    private vendor.mediatek.hardware.mtkpower.V1_0.IMtkPower mOldMtkPerf;
+
     StatusBarManagerInternal getStatusBarManagerInternal() {
         synchronized (mServiceAcquireLock) {
             if (mStatusBarManagerInternal == null) {
@@ -474,6 +477,19 @@ public class DisplayPolicy {
             mScreenOnEarly = true;
             mScreenOnFully = true;
         }
+        try {
+            mMtkPerf = vendor.mediatek.hardware.mtkpower.V1_1.IMtkPerf.getService();
+        } catch(Throwable t) {
+            android.util.Log.d("PHH-Power", "Retrieving mtkpower 1.1", t);
+            mMtkPerf = null;
+        }
+
+        try {
+            mOldMtkPerf = vendor.mediatek.hardware.mtkpower.V1_0.IMtkPower.getService();
+        } catch(Throwable t) {
+            android.util.Log.d("PHH-Power", "Retrieving mtkpower 1.0", t);
+            mOldMtkPerf = null;
+        }
 
         final Looper looper = UiThread.getHandler().getLooper();
         mHandler = new PolicyHandler(looper);
@@ -553,6 +569,17 @@ public class DisplayPolicy {
                             mService.mPowerManagerInternal.setPowerBoost(
                                     Boost.INTERACTION, duration);
                         }
+                        if(mOldMtkPerf != null) {
+                            try {
+                                android.util.Log.d("PHH-Power", "mtk1 fling power hint");
+                                int hint = 36; // MTKPOWER_HINT_APP_TOUCH
+                                if("rotate".equals(SystemProperties.get("persist.sys.phh.touch_hint")))
+                                    hint = 35; // MTKPOWER_HINT_APP_ROTATE
+                                mOldMtkPerf.mtkPowerHint(hint, duration);
+                            } catch(Throwable t) {
+                                android.util.Log.d("PHH-Power", "Failed sending touch power hint", t);
+                            }
+                        }
                     }
 
                     @Override
@@ -571,6 +598,28 @@ public class DisplayPolicy {
                         if (listener != null) {
                             listener.onTouchStart();
                         }
+                        if(mMtkPerf != null) {
+                            try {
+                                android.util.Log.d("PHH-Power", "mtk power hint");
+                                int hint = 25; // MTKPOWER_HINT_APP_TOUCH
+                                if("rotate".equals(SystemProperties.get("persist.sys.phh.touch_hint")))
+                                    hint = 24; // MTKPOWER_HINT_APP_ROTATE
+                                mMtkPerf.perfCusLockHint(hint, 1000);
+                            } catch(Throwable t) {
+                                android.util.Log.d("PHH-Power", "Failed sending touch power hint", t);
+                            }
+                        }
+                        if(mOldMtkPerf != null) {
+                            try {
+                                android.util.Log.d("PHH-Power", "mtk1 power hint");
+                                int hint = 36; // MTKPOWER_HINT_APP_TOUCH
+                                if("rotate".equals(SystemProperties.get("persist.sys.phh.touch_hint")))
+                                    hint = 35; // MTKPOWER_HINT_APP_ROTATE
+                                mOldMtkPerf.mtkPowerHint(hint, 1000);
+                            } catch(Throwable t) {
+                                android.util.Log.d("PHH-Power", "Failed sending touch power hint", t);
+                            }
+                        }
                     }
 
                     @Override
