diff --git a/sdk/src/java/org/lineageos/internal/util/PowerMenuConstants.java b/sdk/src/java/org/lineageos/internal/util/PowerMenuConstants.java
index b04deb2ede31e4f598dce46c94d6a5dc38f7e2e6..4f0b4dad6cae0a018bd85852373dee8bd4982e89 100644
--- a/sdk/src/java/org/lineageos/internal/util/PowerMenuConstants.java
+++ b/sdk/src/java/org/lineageos/internal/util/PowerMenuConstants.java
@@ -22,6 +22,7 @@ public class PowerMenuConstants {
     public static final String GLOBAL_ACTION_KEY_POWER = "power";
     public static final String GLOBAL_ACTION_KEY_RESTART = "restart";
     public static final String GLOBAL_ACTION_KEY_SCREENSHOT = "screenshot";
+    public static final String GLOBAL_ACTION_KEY_ONTHEGO = "onthego";
     public static final String GLOBAL_ACTION_KEY_AIRPLANE = "airplane";
     public static final String GLOBAL_ACTION_KEY_USERS = "users";
     public static final String GLOBAL_ACTION_KEY_SETTINGS = "settings";
@@ -48,6 +49,7 @@ public class PowerMenuConstants {
         GLOBAL_ACTION_KEY_POWER,
         GLOBAL_ACTION_KEY_RESTART,
         GLOBAL_ACTION_KEY_SCREENSHOT,
+        GLOBAL_ACTION_KEY_ONTHEGO,
         GLOBAL_ACTION_KEY_AIRPLANE,
         GLOBAL_ACTION_KEY_USERS,
         GLOBAL_ACTION_KEY_SETTINGS,
