diff --git a/server/Controllers.cpp b/server/Controllers.cpp
index 00ee186d6f72109f3f4e80de076667d09664ce9d..faa6f135597155298aca38797c0e5f4b58a8a3c6 100644
--- a/server/Controllers.cpp
+++ b/server/Controllers.cpp
@@ -284,7 +284,7 @@ void Controllers::init() {
         // As such simply exit netd.  This may crash loop the system, but by failing
         // to bootup we will trigger rollback and thus this offers us protection against
         // a mainline update breaking things.
-        exit(1);
+        // exit(1);
     }
     gLog.info("Enabling bandwidth control: %" PRId64 "us", s.getTimeAndResetUs());
 
diff --git a/server/TetherController.cpp b/server/TetherController.cpp
index 47496fcd9b7150611600dc62a6b5d5a859e81ebd..a03dba454b08f4d14af4296d8da5e272ccaed4cb 100644
--- a/server/TetherController.cpp
+++ b/server/TetherController.cpp
@@ -719,17 +719,23 @@ int TetherController::setForwardRules(bool add, const char *intIface, const char
         "*raw\n"
         "%s %s -i %s -m rpfilter --invert ! -s fe80::/64 -j DROP\n"
         "COMMIT\n", op, LOCAL_RAW_PREROUTING, intIface);
-    if (iptablesRestoreFunction(V6, rpfilterCmd, nullptr) == -1 && add) {
+    if (iptablesRestoreFunction(V6, rpfilterCmd, nullptr) == -1 && add && false) {
         return -EREMOTEIO;
     }
 
-    std::vector<std::string> v4 = {
+    std::vector<std::string> v4Ftp = {
             "*raw",
-            StringPrintf("%s %s -p tcp --dport 21 -i %s -j CT --helper ftp", op,
-                         LOCAL_RAW_PREROUTING, intIface),
+            StringPrintf("%s %s -p tcp --dport 21 -i %s -j CT --helper ftp",
+                            op, LOCAL_RAW_PREROUTING, intIface),
             StringPrintf("%s %s -p tcp --dport 1723 -i %s -j CT --helper pptp", op,
                          LOCAL_RAW_PREROUTING, intIface),
             "COMMIT",
+    };
+    if(iptablesRestoreFunction(V4, Join(v4Ftp, '\n'), nullptr) == -1) {
+            ALOGE("Failed adding iptables CT target on FTP.");
+    }
+
+    std::vector<std::string> v4 = {
             "*filter",
             StringPrintf("%s %s -i %s -o %s -m state --state ESTABLISHED,RELATED -g %s", op,
                          LOCAL_FORWARD, extIface, intIface, LOCAL_TETHER_COUNTERS_CHAIN),
diff --git a/server/main.cpp b/server/main.cpp
index 0e81d4e5c102d488d4a7a2ad78badb066032755a..18287d378f18763066adfbecde12cb3ebd9c8265 100644
--- a/server/main.cpp
+++ b/server/main.cpp
@@ -134,7 +134,7 @@ int main() {
 
     if (libnetd_updatable_init(cg2_path.c_str())) {
         ALOGE("libnetd_updatable_init failed");
-        exit(1);
+        //exit(1);
     }
     gLog.info("libnetd_updatable_init success");
 
